File: CSharp/Supermarket/Model/Core/EAN.cs
using System;
using System.Linq;

namespace Supermarket.Model.Core
{
    /// <summary>
    /// Represents a European Article Number (EAN9 or EAN13).
    /// </summary>
    public class EAN
    {
        private EAN(string code)
        {
            if (string.IsNullOrWhiteSpace(code))
                throw new ArgumentException("The code must not be empty.", nameof(code));
            if (!ContainsJustDigits(code))
                throw new ArgumentException("The code must consist only of digits.", nameof(code));
            if (code.Length != 9 && code.Length != 13)
                throw new ArgumentException("The code must be 9 or 13 characters long.", nameof(code));
            Code = code;
        }

        public string Code { get; }

        /// <summary>
        /// Creates a EAN object based on the code passed. The validity of the code is not checked.
        /// </summary>
        /// <param name="code"></param>
        /// <returns></returns>
        public static EAN NewEAN(string code)
        {
            return new EAN(code);
        }

        private static bool ContainsJustDigits(string text)
        {
            return text.All(c => c >= '0' && c <= '9');
        }

        public override string ToString() => $"EAN{Code.Length}: {Code}";

        #region Equality Members

        protected bool Equals(EAN other)
        {
            return string.Equals(Code, other.Code);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((EAN) obj);
        }

        public override int GetHashCode()
        {
            return Code.GetHashCode();
        }

        public static bool operator ==(EAN left, EAN right)
        {
            return Equals(left, right);
        }

        public static bool operator !=(EAN left, EAN right)
        {
            return !Equals(left, right);
        }

        #endregion
    }
}

File: java/src/main/java/supermarket/model/core/EAN.java
package supermarket.model.core;

public class EAN {

	private String code;

	public EAN(String code) {
		if (isNullOrWhiteSpace(code))
			throw new IllegalArgumentException("The code must not be empty.");
		if (!containsJustDigits(code))
			throw new IllegalArgumentException("The code must consist only if digits but was "+ code);
		if (code.length() != 9 && code.length() != 13)
			throw new IllegalArgumentException("The code must be 9 or 13 characters long but was " + code);

		this.code = code;
	}

	private boolean isNullOrWhiteSpace(String text) {
		return text == null || text.isEmpty();
	}
	private static boolean containsJustDigits(String text) {
		return text.matches("[0-9]*");
	}

	public String getCode() {
		return code;
	}

	public static EAN newEAN(String code) {
		return new EAN(code);
	}


	@Override
	public String toString() {
		return String.format("EAN%i: %s", code.length(), code);
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((code == null) ? 0 : code.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		EAN other = (EAN) obj;
		if (code == null) {
			if (other.code != null)
				return false;
		} else if (!code.equals(other.code))
			return false;
		return true;
	}

}
